File: SiteScrapers/scrapers/AkkelisAudio.ts
import { BaseScraper } from '../BaseScraper';
import { ListingResult, ScraperOptions } from '../types';
import { extractPrice, normalizeUrl, filterByPrice } from '../utils';

/**
 * Playwright-based scraper for Akkelis Audio
 * Scrapes the Fyndhörnan (Finding Corner) category page
 *
 * Base URL: https://www.akkelisaudio.com
 * Type: HTML-based category scraper
 * Features: html_parsing, price_extraction, image_handling
 */
export class AkkelisAudioPlaywright extends BaseScraper {
  private readonly categoryUrl: string;

  constructor(options?: ScraperOptions) {
    super('https://www.akkelisaudio.com', 'Akkelis Audio', options);
    this.categoryUrl = `${this.baseUrl}/fyndhornan/`;
  }

  async search(
    query: string,
    minPrice?: number,
    maxPrice?: number
  ): Promise<ListingResult[]> {
    if (!query?.trim()) {
      return [];
    }

    const page = await this.newPage();
    try {
      await this.rateLimit();

      // Navigate to category page
      await page.goto(this.categoryUrl, {
        waitUntil: 'domcontentloaded',
        timeout: this.options.timeout,
      });

      // Wait for items to appear
      await page.waitForSelector('.tws-list--grid-item', {
        timeout: 3000,
      });

      // Extract all items
      const items = await page.locator('.tws-list--grid-item').all();
      const results: ListingResult[] = [];
      const queryLower = query.toLowerCase();

      for (const item of items) {
        // Extract title
        const titleElement = item.locator('.tws-util-heading--heading a');
        const title = await titleElement.textContent();

        if (!title?.toLowerCase().includes(queryLower)) {
          continue;
        }

        // Extract URL
        const url = await titleElement.getAttribute('href');

        // Extract price
        const priceElement = item.locator(
          '.tws-api--price-current, .tws-api--price-regular'
        );
        const priceText = await priceElement.textContent();
        const price = extractPrice(priceText || '');

        // Extract description
        const descElement = item.locator('.tws-article-labels--label-text');
        const description = await descElement.textContent();

        // Extract image
        const imageElement = item.locator('.tws-img');
        const imageUrl = await imageElement.getAttribute('source');

        const listing: ListingResult = {
          title: title?.trim() || '',
          url: normalizeUrl(url || '', this.baseUrl),
          price,
          description: description?.trim(),
          imageUrl: imageUrl || undefined,
          location: undefined,
          postedDate: undefined,
          rawData: { source: 'akkelis' },
        };

        results.push(listing);
      }

      // Filter by price
      return filterByPrice(results, minPrice, maxPrice);
    } finally {
      await page.close();
    }
  }
}

export function createAkkelisAudioPlaywright(
  options?: ScraperOptions
): AkkelisAudioPlaywright {
  return new AkkelisAudioPlaywright(options);
}

File: Scrapers/akkelisaudio.py
import asyncio
from typing import List, Optional

from bs4 import BeautifulSoup

from base import BaseScraper, ListingResult


class AkkelisAudioScraper(BaseScraper):
    """Scrapes Akkelis Audio's Fyndhörnan listing."""

    def __init__(self):
        super().__init__("https://www.akkelisaudio.com", "Akkelis Audio")
        self.category_url = "https://www.akkelisaudio.com/fyndhornan/"

    def _search_sync(self, query: str, min_price: Optional[float], max_price: Optional[float]) -> List[ListingResult]:
        query_lower = (query or "").lower()
        if not query_lower:
            return []

        response = self.session.get(self.category_url, timeout=30)
        response.raise_for_status()
        soup = BeautifulSoup(response.text, "html.parser")
        items = soup.select(".tws-list--grid-item")
        results: List[ListingResult] = []

        for item in items:
            title_tag = item.select_one(".tws-util-heading--heading a")
            if not title_tag:
                continue
            title = title_tag.get_text(strip=True)
            if query_lower not in title.lower():
                continue

            link = title_tag.get("href")
            price_current = item.select_one(".tws-api--price-current") or item.select_one(".tws-api--price-regular")
            price_text = price_current.get_text(strip=True) if price_current else None
            description = item.select_one(".tws-article-labels--label-text")
            image = item.select_one(".tws-img")
            image_url = image.get("source") if image else None

            listing = ListingResult(
                title=title,
                description=description.get_text(strip=True) if description else None,
                price=self._extract_price(price_text) if price_text else None,
                url=link,
                image_url=image_url,
                posted_date=None,
                location=None,
                raw_data={"source": "akkelis"},
            )

            if min_price and listing.price and listing.price < min_price:
                continue
            if max_price and listing.price and listing.price > max_price:
                continue
            results.append(listing)

        return results

    async def search(
        self, query: str, min_price: Optional[float] = None, max_price: Optional[float] = None, **kwargs
    ) -> List[ListingResult]:
        return await asyncio.to_thread(self._search_sync, query, min_price, max_price)
